From 355e834a05c24fbdb7e572eec47ce1c38cbe7cb5 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 5 Sep 2015 09:12:04 -0400
Subject: [PATCH] Allow commands to dynamically decide if they should be
 handled or passed upstream


diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/CommandBypassException.java b/api/src/main/java/net/md_5/bungee/api/plugin/CommandBypassException.java
new file mode 100644
index 00000000..0c3419c8
--- /dev/null
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/CommandBypassException.java
@@ -0,0 +1,8 @@
+package net.md_5.bungee.api.plugin;
+
+/**
+ * Thrown from inside a command to tell the proxy to pass the command upstream
+ */
+public class CommandBypassException extends RuntimeException {
+
+}
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index 71a5a158..ed8d0f6e 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -174,6 +174,8 @@ public class PluginManager
                     tabResults.add( s );
                 }
             }
+        } catch ( CommandBypassException ex ) {
+            return false;
         } catch ( Exception ex )
         {
             sender.sendMessage( ChatColor.RED + "An internal error occurred whilst executing this command, please check the console log for details." );
-- 
2.11.0 (Apple Git-81)

