From 655539c6c02653d93c874621929e0e33aff65875 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 10 Oct 2015 18:38:52 -0400
Subject: [PATCH] Register PluginLoggers with the LogManager


diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java b/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java
index bc91208a..d9cc5870 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java
@@ -85,7 +85,7 @@ public class Plugin
         this.proxy = proxy;
         this.description = description;
         this.file = description.getFile();
-        this.logger = new PluginLogger( this );
+        this.logger = PluginLogger.get( this );
     }
 
     //
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginLogger.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginLogger.java
index b304ec0d..8b52dd8d 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginLogger.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginLogger.java
@@ -1,10 +1,29 @@
 package net.md_5.bungee.api.plugin;
 
+import java.util.logging.LogManager;
 import java.util.logging.LogRecord;
 import java.util.logging.Logger;
 
 public class PluginLogger extends Logger
 {
+    public static PluginLogger get(Plugin context) {
+        LogManager lm = LogManager.getLogManager();
+        Logger logger = lm.getLogger(context.getClass().getCanonicalName());
+
+        if(logger instanceof PluginLogger) {
+            return (PluginLogger) logger;
+        } else {
+            PluginLogger pluginLogger = new PluginLogger(context);
+
+            // Register the logger under the plugin's name, unless some other logger is already using the name
+            if(logger == null) {
+                lm.addLogger(pluginLogger);
+                pluginLogger.setParent(context.getProxy().getLogger()); // addLogger changes this, change it back
+            }
+
+            return pluginLogger;
+        }
+    }
 
     private final String pluginName;
 
diff --git a/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java b/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java
index 75e71c4c..3647747a 100644
--- a/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java
+++ b/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java
@@ -31,7 +31,6 @@ public class BungeeLogger extends Logger
             addHandler( fileHandler );
 
             ColouredWriter consoleHandler = new ColouredWriter( reader );
-            consoleHandler.setLevel( Level.INFO );
             consoleHandler.setFormatter( formatter );
             addHandler( consoleHandler );
         } catch ( IOException ex )
-- 
2.11.0 (Apple Git-81)

